vm/systemSmartContracts/governance.go

```diff
@@ -236,230 +229,470 @@ func (g *governanceContract) whiteListProposal(args *vmcommon.ContractCallInput)
+	stakeLockKey := append([]byte(stakeLockPrefix), address...)
+	lastData := g.eei.GetStorage(stakeLockKey)
+	lastEndNonce := uint64(0)
+	if len(lastData) > 0 {
+		lastEndNonce = big.NewInt(0).SetBytes(lastData).Uint64()
 	}
 
-	generalProposal := &GeneralProposal{}
-	err := g.marshalizer.Unmarshal(generalProposal, marshaledData)
-	if err != nil {
-		return nil, err
+	if lastEndNonce < endNonce {
+		g.eei.SetStorage(stakeLockKey, big.NewInt(0).SetUint64(endNonce).Bytes())
 	}
-
-	return generalProposal, nil
 }
 
-func (g *governanceContract) isWhiteListed(address []byte) bool {
-	key := append([]byte(whiteListPrefix), address...)
-	marshaledData := g.eei.GetStorage(key)
-	if len(marshaledData) == 0 {
-		return false
-	}
-
-	key = append([]byte(proposalPrefix), address...)
-	marshaledData = g.eei.GetStorage(key)
-	generalProposal := &GeneralProposal{}
-	err := g.marshalizer.Unmarshal(generalProposal, marshaledData)
-	if err != nil {
+func isStakeLocked(eei vm.SystemEI, governanceAddress []byte, address []byte) bool {
+	stakeLockKey := append([]byte(stakeLockPrefix), address...)
+	lastData := eei.GetStorageFromAddress(governanceAddress, stakeLockKey)
+	if len(lastData) == 0 {
 		return false
 	}
 
-	return generalProposal.Voted
+	lastEndNonce := big.NewInt(0).SetBytes(lastData).Uint64()
+	return eei.BlockChainHook().CurrentNonce() < lastEndNonce
 }
 
-func (g *governanceContract) whiteListAtGenesis(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+// delegateVote casts a vote from a validator run by WASM SC and delegates it to some. This function receives 4 parameters:
+//  args.Arguments[0] - proposal reference (github commit)
+//  args.Arguments[1] - vote option (yes, no, veto)
+//  args.Arguments[2] - delegatedTo
+//  args.Arguments[3] - balance to vote
+func (g *governanceContract) delegateVote(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.Arguments) != 4 {
+		g.eei.AddReturnMessage("invalid number of arguments")
+		return vmcommon.UserError
+	}
 	if args.CallValue.Cmp(zero) != 0 {
-		log.Warn("whiteList at genesis should be without callValue")
+		g.eei.AddReturnMessage("function is not payable")
 		return vmcommon.UserError
 	}
-	if g.isWhiteListed(args.CallerAddr) {
-		log.Warn("address is already whiteListed")
+	err := g.eei.UseGas(g.gasCost.MetaChainSystemSCsCost.DelegateVote)
+	if err != nil {
+		g.eei.AddReturnMessage("not enough gas")
+		return vmcommon.OutOfGas
+	}
+	if !core.IsSmartContractAddress(args.CallerAddr) {
+		g.eei.AddReturnMessage("only SC can call this")
 		return vmcommon.UserError
 	}
-	if len(args.Arguments) != 0 {
-		log.Warn("excepted argument number is 0")
+	if len(args.Arguments[3]) != len(args.CallerAddr) {
+		g.eei.AddReturnMessage("invalid delegator address")
 		return vmcommon.UserError
 	}
-	if g.proposalExists(args.CallerAddr) {
-		log.Warn("proposal with this key already exists")
+
+	voterAddress := args.CallerAddr
```

Comment:
is this correct? The voter is the same as the caller address which is a SC address? In case of yes, what happens if the SC is constructed in such a way it distorts the vote cast by the delegator?

Reply:
it is correct. Voter is actually the SC, the information of delegation is Under VoteDetails.DelegatedTo

Reply:
Then the code inside the SC can cast any kind of vote. I think we should change this model.

Reply:
There is no way to reason about data coming from another shard. you do not know whether it is a real user which voted or not. People will not blindly delegate to SCs. And we can simply make a post about a SC which makes wrong votes.

Reply:
ok, makes sense